### ISSUE TITLE
Exclusion filter for volumes iffy

### ISSUE THREAD

**Issue by ste-fan** [digint/btrbk#291]
Consider this minimum `btrbk.conf` for backup of local `@home` subvol on a laptop to an external backup drive, whose `@data` subvol is also snapshotted, but not backed up.

```
volume /mnt/btrfs/HOME
    target /mnt/btrfs/BACKUP/laptop
    subvolume @home

volume /mnt/btrfs/BACKUP
    subvolume @data
```

To trigger an automatic `resume` operation after the external disk is plugged in I want to select all subvolumes that back up to any subdir on the external disk, but exclude subvolumes on the external drive itself (which are also matched by the general filter).

```shellsession
$ btrbk --config=btrbk.conf --exclude=/mnt/btrfs/BACKUP list config /mnt/btrfs/BACKUP/\*
source_subvol          snapshot_path    snapshot_name  target_path
---------------------  ---------------  -------------  ------------------------
/mnt/btrfs/HOME/@home  /mnt/btrfs/HOME  @home          /mnt/btrfs/BACKUP/laptop
```

However in an action `<command>` context the exclusion doesn't take full effect. At least in the output the to be excluded subvolume is still mentioned. 
```shellsession
# btrbk --config=btrbk.conf --exclude=/mnt/btrfs/BACKUP dryrun /mnt/btrfs/BACKUP/\*
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.28.2)

    Date:   Mon Aug 26 08:14:06 2019
    Config: btrbk.conf
    Dryrun: YES
    Filter: /mnt/btrfs/BACKUP/*

Options:
    exclude: /mnt/btrfs/BACKUP

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/mnt/btrfs/HOME/@home
+++ /mnt/btrfs/HOME/@home.20190826
>>> /mnt/btrfs/BACKUP/laptop/@home.20190826

/mnt/btrfs/BACKUP/@data
<no_action>

NOTE: Dryrun was active, none of the operations above were actually executed!
```
(see `exlude_vol.debug` for debug output of `--dry-run --preserve snapshot` instead of `dryrun`)

If I exclude the subvolume directly (their names always start with `@` in my case) it will not be displayed.
```shellsession
# btrbk --config=btrbk.conf --exclude=/mnt/btrfs/BACKUP/@\* dryrun /mnt/btrfs/BACKUP/\*
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.28.2)

    Date:   Mon Aug 26 08:15:19 2019
    Config: btrbk.conf
    Dryrun: YES
    Filter: /mnt/btrfs/BACKUP/*

Options:
    exclude: /mnt/btrfs/BACKUP/@*

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/mnt/btrfs/HOME/@home
+++ /mnt/btrfs/HOME/@home.20190826
>>> /mnt/btrfs/BACKUP/laptop/@home.20190826

NOTE: Dryrun was active, none of the operations above were actually executed!
```
(see `exlude_subvols.debug` for debug output of `--dry-run --preserve snapshot` instead of `dryrun`)

A diff of the two debug outputs doesn't tell me why the two outputs should differ. There should be no difference between excluding a volume and excluding all of its subvolumes, right?
```diff
--- exclude_vol.debug	2019-08-26 08:24:32.990246845 +0200
+++ exclude_subvols.debug	2019-08-26 08:24:32.158237938 +0200
-Skipping volume "/mnt/btrfs/BACKUP": command line argument "--exclude=/mnt/btrfs/BACKUP"
+Skipping subvolume "/mnt/btrfs/BACKUP/@data": command line argument "--exclude=/mnt/btrfs/BACKUP/@*"
+Skipping volume "/mnt/btrfs/BACKUP": All subvolumes excluded
```

I have not yet seen a run where anything else beside the `<no_action>` mention happened to the subvolumes of an excluded volume. However it seems weird to me and for now  in cronjobs I use the `@*` exclusion filter that works on subvolumes to be on the safe side.

---

<details>
<summary><code>exclude_vol.debug</code> (debug output of excluding the external volume)</summary>
<p>

```
btrbk command line client, version 0.28.2  (Mon Aug 26 08:21:36 2019)
Using configuration: btrbk.conf
Skipping volume "/mnt/btrfs/BACKUP": command line argument "--exclude=/mnt/btrfs/BACKUP"
Initializing volume section: /mnt/btrfs/HOME
Resolving btrfs mount point for: /mnt/btrfs/HOME
### readlink -e /mnt/btrfs/HOME
Command execution successful
Real path for "/mnt/btrfs/HOME" is: /mnt/btrfs/HOME
### cat /proc/self/mountinfo
Command execution successful
Btrfs mount point for "/mnt/btrfs/HOME": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
### btrfs subvolume list -a -c -u -q -R /mnt/btrfs/HOME
Command execution successful
Parsed 32 total subvolumes for filesystem at: /mnt/btrfs/HOME
### btrfs subvolume list -a -r /mnt/btrfs/HOME
Command execution successful
Parsed 31 readonly subvolumes for filesystem at: /mnt/btrfs/HOME
### btrfs subvolume show --rootid=5 /mnt/btrfs/HOME
Command execution successful
Parsed 11 subvolume detail items: /mnt/btrfs/HOME
No UUID on btrfs root (id=5): /mnt/btrfs/HOME
found btrfs root: /mnt/btrfs/HOME
Initializing subvolume section: /mnt/btrfs/HOME/@home
Resolving btrfs mount point for: /mnt/btrfs/HOME/@home
### readlink -e /mnt/btrfs/HOME/@home
Command execution successful
Real path for "/mnt/btrfs/HOME/@home" is: /mnt/btrfs/HOME/@home
Btrfs mount point for "/mnt/btrfs/HOME/@home": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
Found "/mnt/btrfs/HOME/@home" (id=257) in btrfs tree of: /mnt/btrfs/HOME
Resolving btrfs mount point for: /mnt/btrfs/HOME
Btrfs mount point for "/mnt/btrfs/HOME": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
Skipping target tree readin (preserving all snapshots and backups)
Evaluating snapshot creation for: /mnt/btrfs/HOME/@home
Snapshot creation enabled (snapshot_create=always)
Creating subvolume snapshot for: /mnt/btrfs/HOME/@home
[snapshot] source: /mnt/btrfs/HOME/@home
[snapshot] target: /mnt/btrfs/HOME/@home.20190826
### (dryrun) btrfs subvolume snapshot -r /mnt/btrfs/HOME/@home /mnt/btrfs/HOME/@home.20190826
Skipping backup creation (btrbk snapshot)
Preserving all snapshots and backups
Completed within: 0s  (Mon Aug 26 08:21:36 2019)
```
</p>
</details>

<details>
<summary><code>exclude_subvols.debug</code> (debug output of excluding all of its subvolumes)</summary>
<p>

```
btrbk command line client, version 0.28.2  (Mon Aug 26 08:21:51 2019)
Using configuration: btrbk.conf
Skipping subvolume "/mnt/btrfs/BACKUP/@data": command line argument "--exclude=/mnt/btrfs/BACKUP/@*"
Skipping volume "/mnt/btrfs/BACKUP": All subvolumes excluded
Initializing volume section: /mnt/btrfs/HOME
Resolving btrfs mount point for: /mnt/btrfs/HOME
### readlink -e /mnt/btrfs/HOME
Command execution successful
Real path for "/mnt/btrfs/HOME" is: /mnt/btrfs/HOME
### cat /proc/self/mountinfo
Command execution successful
Btrfs mount point for "/mnt/btrfs/HOME": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
### btrfs subvolume list -a -c -u -q -R /mnt/btrfs/HOME
Command execution successful
Parsed 32 total subvolumes for filesystem at: /mnt/btrfs/HOME
### btrfs subvolume list -a -r /mnt/btrfs/HOME
Command execution successful
Parsed 31 readonly subvolumes for filesystem at: /mnt/btrfs/HOME
### btrfs subvolume show --rootid=5 /mnt/btrfs/HOME
Command execution successful
Parsed 11 subvolume detail items: /mnt/btrfs/HOME
No UUID on btrfs root (id=5): /mnt/btrfs/HOME
found btrfs root: /mnt/btrfs/HOME
Initializing subvolume section: /mnt/btrfs/HOME/@home
Resolving btrfs mount point for: /mnt/btrfs/HOME/@home
### readlink -e /mnt/btrfs/HOME/@home
Command execution successful
Real path for "/mnt/btrfs/HOME/@home" is: /mnt/btrfs/HOME/@home
Btrfs mount point for "/mnt/btrfs/HOME/@home": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
Found "/mnt/btrfs/HOME/@home" (id=257) in btrfs tree of: /mnt/btrfs/HOME
Resolving btrfs mount point for: /mnt/btrfs/HOME
Btrfs mount point for "/mnt/btrfs/HOME": /mnt/btrfs/HOME (mount_source=/dev/mapper/home_crypt, subvolid=5)
Skipping target tree readin (preserving all snapshots and backups)
Evaluating snapshot creation for: /mnt/btrfs/HOME/@home
Snapshot creation enabled (snapshot_create=always)
Creating subvolume snapshot for: /mnt/btrfs/HOME/@home
[snapshot] source: /mnt/btrfs/HOME/@home
[snapshot] target: /mnt/btrfs/HOME/@home.20190826
### (dryrun) btrfs subvolume snapshot -r /mnt/btrfs/HOME/@home /mnt/btrfs/HOME/@home.20190826
Skipping backup creation (btrbk snapshot)
Preserving all snapshots and backups
Completed within: 0s  (Mon Aug 26 08:21:51 2019)
```
</p>
</details>

**Comment by digint** [comment]
Thanks for reporting!

Fixed in 8570ee585d38248c41a1803e7bf00a5b18c0e455

Note that the bug was only on the backup summary, no functional changes.

**Comment by ste-fan** [comment]
Works like a charm. Thanks for clearing up that this was merely a summary output issue.

```shellsession
# btrbk --config=btrbk.conf --exclude=/mnt/btrfs/BACKUP dryrun /mnt/btrfs/BACKUP/\*
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.29.0-dev)

    Date:   Mon Sep  9 09:13:04 2019
    Config: btrbk.conf
    Dryrun: YES
    Filter: /mnt/btrfs/BACKUP/*

Options:
    exclude: /mnt/btrfs/BACKUP

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/mnt/btrfs/HOME/@home
+++ /mnt/btrfs/HOME/@home.20190909
>>> /mnt/btrfs/BACKUP/laptop/@home.20190909

NOTE: Dryrun was active, none of the operations above were actually executed!
```

**Comment by digint** [comment]
included in [btrbk-0.29.0](/digint/btrbk/releases/tag/v0.29.0), thanks!
